MIPs/mip-x.md

```diff
@@ -0,0 +1,140 @@
+---
+MIP: X
+Title: Implement JSON-RPC Request Batching
+Status: Draft
+discussions-to: https://github.com/MetaMask/metamask-improvement-proposals/discussions
+Author(s): Jiexi Luan (@jiexi)
+Type: Maintainer
+Created: 2024-03-25
+---
+
+## Summary
+This proposal aims to extend the MetaMask JSON-RPC API that is accessible via `window.ethereum.request()` to include support for batching requests. Batched requests would be processed in parallel by the wallet. All passthrough requests in a batch will be sent to the RPC endpoint in a batch. All requests requiring user confirmation will have results returned after user approval/rejection. Results will be returned as available and are not guaranteed to be in the same order as the respective requests. Introducing native support for request batching allows developers to improve Dapp efficiency and aligns the MetaMask JSON-RPC API closer to the [JSON-RPC 2.0 specification](https://www.jsonrpc.org/specification#batch).
+
+## Motivation
+
+
+# Usage Example
+```js
+// Request
+await window.ethereum.request([
+    {
+    "id": 1,
```

Comment:
currently the `id` gets generated on metamasks side, for batching we should probably honor their `id`

Reply:
It seems that the generated id gets mapped back to any user passed one

https://github.com/MetaMask/core/blob/main/packages/json-rpc-engine/src/idRemapMiddleware.ts

Reply:
to clarify, i think we need to make changes to the provider/middleware to not add an id to a request object in the case of batches